lib/shared-types.d.ts

```diff
@@ -0,0 +1,293 @@
+      | {[code: number]: Construct | Construct[]},
+    ok: Okay,
+    nok?: NotOkay
+  ) => (code: number) => void
+
+  check: (
+    constructInfo:
+      | Construct
+      | Construct[]
+      | {[code: number]: Construct | Construct[]},
+    ok: Okay,
+    nok?: NotOkay
+  ) => (code: number) => void
+
+  /**
+   * Lazy is used for lines that were not properly preceded by the container.
+   */
+  lazy: (
+    constructInfo:
+      | Construct
+      | Construct[]
+      | {[code: number]: Construct | Construct[]},
+    ok: Okay,
+    nok?: NotOkay
+  ) => void
+}
+
+/**
+ * A state function should return another function: the next state-as-a-function to go to.
+ *
+ * But there is one case where they return void: for the eof character code (at the end of a value)
+ * The reason being: well, there isn’t any state that makes sense, so void works well. Practically
+ * that has also helped: if for some reason it was a mistake, then an exception is throw because
+ * there is no next function, meaning it surfaces early.
+ */
+export type State = (code: number) => State | void
+
+/**
+ *
+ */
+export type Okay = State
+
+/**
+ *
+ */
+export type NotOkay = State
+
+/**
+ *
+ */
+export interface Tokenizer {
+  previous: Code
+  events: Event[]
+  parser: Parser
+  sliceStream: (token: Token) => Chunk[]
+  sliceSerialize: (token: Token) => string
+  now: () => Point
+  defineSkip: (value: Point) => void
+  write: (slice: Chunk[]) => Event[]
+}
+
+export type Resolve = (events: Event[], context: Tokenizer) => Event[]
+
+export type Tokenize = (context: Tokenizer, effects: Effects) => State
+
+export interface Construct {
+  tokenize: Tokenize
+  partial?: boolean
+  resolve?: Resolve
+  resolveTo?: Resolve
+  resolveAll?: Resolve
+  concrete?: boolean
+  interruptible?: boolean
+  lazy?: boolean
+}
+
+/**
+ *
+ */
+export interface Parser {
+  constructs: Record<NonNullable<Code> | 'null', Construct | Construct[]>
+  content: (from: Point) => Tokenizer
+  document: (from: Point) => Tokenizer
+  flow: (from: Point) => Tokenizer
+  string: (from: Point) => Tokenizer
+  text: (from: Point) => Tokenizer
+  defined: string[]
+}
+
+/**
+ *
+ */
+export interface TokenizerThis {
+  events: Event[]
+  interrupt: unknown
+  lazy: unknown
+  containerState: {
+    marker: number
+    type: Type
+    initialBlankLine: unknown
+    size: number
+    _closeFlow: unknown
+    furtherBlankLines: unknown
+  }
+}
+
+/**
+ *
+ */
+export type Compile = (slice: Event[]) => string
+
+/**
+ * https://github.com/micromark/micromark#syntaxextension
+ */
+export interface SyntaxExtension {
+  content: string
+  document: unknown
+  flow: unknown
+  string: unknown
+  text: unknown
+}
+
+/**
+ * https://github.com/micromark/micromark#htmlextension
+ */
+export type HtmlExtension =
+  | {
+      enter: unknown
+    }
+  | {
+      exit: unknown
+    }
+
+export type ParseOptions = {
+  // Array of syntax extensions
+  //
+  extensions?: SyntaxExtension[]
+} & CompileOptions
```

Comment:
Should there be a `ParseAndCompileOptions` that includes both? Or just one called `Options`? So `Options` are given by users to buffer and stream, but `lib/parse` only takes the parse options, and `lib/compiler/html` takes the compile options?

Reply:
Check if I'm right:

We have three kinds of options:
- for `lib/parse`, we have `ParseOptions`, which has one attribute:
  - extensions
- for `lib/compile/html`, we have `CompileOptions`, which has four attributes:
  - defaultLineEnding
  - allowDangerousHtml
  - allowDangerousProtocol
  - htmlExtensions
- for `lib/stream` and `lib/index`, we have `Options`, which has all five attributes above

Reply:
Yes, perfect!